analyzers/src/SonarAnalyzer.CSharp/Rules/GenericTypeParametersRequired.cs

```diff
@@ -31,35 +31,50 @@ public sealed class GenericTypeParametersRequired : SonarDiagnosticAnalyzer
 
         protected override void Initialize(SonarAnalysisContext context) =>
             context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                var methodDeclaration = (MethodDeclarationSyntax)c.Node;
+
+                var typeParameters = methodDeclaration
+                        .TypeParameterList
+                        ?.Parameters
+                        .Select(p => c.SemanticModel.GetDeclaredSymbol(p));
+
+                if (typeParameters == null)
                 {
-                    var methodDeclaration = (MethodDeclarationSyntax)c.Node;
+                    return;
+                }
 
-                    var typeParameters = methodDeclaration
-                            .TypeParameterList
-                            ?.Parameters
-                            .Select(p => c.SemanticModel.GetDeclaredSymbol(p));
+                var argumentTypes = methodDeclaration
+                        .ParameterList
+                        .Parameters
+                        .Select(p => c.SemanticModel.GetDeclaredSymbol(p)?.Type);
 
-                    if (typeParameters == null)
-                    {
-                        return;
-                    }
+                var constraintTypes = methodDeclaration
+                    .ConstraintClauses
+                    .OfType<TypeParameterConstraintClauseSyntax>()
+                    .SelectMany(p => p
+                                    .Constraints
+                                    .OfType<TypeConstraintSyntax>()
+                                    .Select(t => c.SemanticModel.GetSymbolInfo(t.Type).Symbol))
+                    .OfType<ITypeSymbol>();
 
-                    var argumentTypes = methodDeclaration
-                            .ParameterList
-                            .Parameters
-                            .Select(p => c.SemanticModel.GetDeclaredSymbol(p)?.Type);
+                var returnType = (ITypeSymbol)c.SemanticModel.GetSymbolInfo(methodDeclaration.ReturnType).Symbol;
```

Comment:
rather than doing a `SymbolInfo` query, can you use the containing type given this callback is only on a `MethodDeclaration`?

```
((IMethodSymbol)c.ContainingSymbol).ReturnType
```